### ISSUE TITLE
Keyboard but other Languages

### ISSUE THREAD

**Issue by Psychpsyo** [Yellow-Dog-Man/Resonite-Issues#985]
### Is your feature request related to a problem? Please describe.

I'm always frustrated when trying to type words on the keyboard because it does not work in my language. Also, the keys are in unfamiliar places.

### Describe the solution you'd like

Implement localized keyboard layouts such as German please.

### Describe alternatives you've considered

Custom keyboards exist but they are a hassle to install and I need to trust the person making it with my userspace.

### Additional Context

This is probably a much bigger issues with languages such as Russian, Korean or Japanese, which makes up quite a lot of the userbase.

**Comment by Frooxius** [comment]
We probably don't want to implement the individual layouts for different languages ourselves - there's a lot of them and we are not familiar with those languages and layouts to implement them well.

Instead, we should provide a mechanism to load existing layouts from external sources or provide a good system for specifying layouts.

I haven't done much research into this yet, so if there are some suggestions on how to approach this technically, they'd be welcome.

I know that some languages like Japanese for example use more advanced methods of input (IME), which will probably require more code support - I don't know how generalized this is or if there are good libraries that would help with this.

**Comment by Psychpsyo** [comment]
> We probably don't want to implement the individual layouts for different languages ourselves - there's a lot of them and we are not familiar with those languages and layouts to implement them well.

I feel like it'd make sense for each Resonite language option to come with a keyboard layout (or layouts?) for that language.
These could be community-sourced just like the locale files themselves.

> Instead, we should provide a mechanism to load existing layouts from external sources or provide a good system for specifying layouts.
> 
> I haven't done much research into this yet, so if there are some suggestions on how to approach this technically, they'd be welcome.

There is existing file formats that can store a keyboard layout though these come with a lot of complications stemming from how physical keyboards work.
Being able to use these might make it simpler to just drop in any existing layout but supporting and editing them would be more difficult over some form of simplified Resonite-specific layout format.

> I know that some languages like Japanese for example use more advanced methods of input (IME), which will probably require more code support - I don't know how generalized this is or if there are good libraries that would help with this.

Those would probably need manual implementation work and I think Korean IMEs are fundamentally different from Japanese / Chinese IMEs which are more similar to the auto-complete feature on phone keyboards. (the same UI could probably be re-used to provide auto-complete suggestions in other languages at some point as well)

Though most languages only require different symbol to key mappings and dead keys. 
A dead key is a key that doesn't type anything on its own, but if followed by another key, types a special version of that character, similar to shift on a phone.
I have a ``´`` dead key for example that I can use to type á by typing ``´`` and then ``A`` (or ``Shift``+``A`` for Á)

**Comment by Frooxius** [comment]
I don't think we should tie it into the locale system, there are layouts that don't necessarily match a particular language and can be used across multiple.

Plus if you're multilingual, you often use layouts that differ from your UI language.

Keeping it separate would be cleanest, we just need a system that can handle these.


I understand number of these principles (e.g. My native language uses the dead keys). What I'm mainly asking for are specific libraries and/or formats that we can build this around, rather than designing and implementing everything from scratch.

That would make this task much smaller in scope and much more likely to get prioritized.

**Comment by shadowpanther** [comment]
> I don't think we should tie it into the locale system, there are layouts that don't necessarily match a particular language and can be used across multiple.

I think what @Psychpsyo suggested is that the layouts themselves could be created as a part of the localisation effort.

I would add to this that after the layouts are created, users should be able to select the ones they use and switch between layouts while using the keyboard. E.g., I would select the English(QWERTY) and the Russian(ЙЦУКЕН) layouts. A German person would probably use the English(QWERTZ) and the German(QWERTZ) ones. For a reference, look into how OSes allow you to select a set of layouts you use from a large assortment and switch between them.

I already have a custom keyboard for exactly my use-case with a layout switch, but it is not versatile - you can't easily add another layout to it.
